crates/ra_ide/src/ssr.rs

```diff
@@ -109,340 +168,863 @@ impl FromStr for SsrQuery {
-        };
-        let match_ = check_opt_nodes(pattern_name, code.name_ref(), placeholders, match_)?;
-        let match_ =
-            check_opt_nodes(pattern_type_args, code.type_arg_list(), placeholders, match_)?;
-        let pattern_args = pattern.syntax().children().find_map(ArgList::cast)?.args();
-        let code_args = code.syntax().children().find_map(ArgList::cast)?.args();
-        let code_args = once(code.expr()?).chain(code_args);
-        check_iter(pattern_args, code_args, placeholders, match_)
-    }
-
-    fn check_method_call_and_call(
-        pattern: MethodCallExpr,
-        code: CallExpr,
-        placeholders: &[Var],
-        match_: Match,
-    ) -> Option<Match> {
-        let (code_name, code_type_args) = if let Some(Expr::PathExpr(path_exr)) = code.expr() {
-            let segment = path_exr.path().and_then(|p| p.segment());
-            (segment.as_ref().and_then(|s| s.name_ref()), segment.and_then(|s| s.type_arg_list()))
+    fn is_keyword(&self, kw: &str) -> bool {
+        if let Some(PatternElement::Token(token)) = self.tokens.get(self.position) {
+            token.text == kw
         } else {
-            (None, None)
-        };
-        let match_ = check_opt_nodes(pattern.name_ref(), code_name, placeholders, match_)?;
-        let match_ =
-            check_opt_nodes(pattern.type_arg_list(), code_type_args, placeholders, match_)?;
-        let code_args = code.syntax().children().find_map(ArgList::cast)?.args();
-        let pattern_args = pattern.syntax().children().find_map(ArgList::cast)?.args();
-        let pattern_args = once(pattern.expr()?).chain(pattern_args);
-        check_iter(pattern_args, code_args, placeholders, match_)
-    }
-
-    fn check_opt_nodes(
-        pattern: Option<impl AstNode>,
-        code: Option<impl AstNode>,
-        placeholders: &[Var],
-        match_: Match,
-    ) -> Option<Match> {
-        match (pattern, code) {
-            (Some(pattern), Some(code)) => check(
-                &pattern.syntax().clone().into(),
-                &code.syntax().clone().into(),
-                placeholders,
-                match_,
-            ),
-            (None, None) => Some(match_),
-            _ => None,
+            false
         }
     }
+}
 
-    fn check_iter<T, I1, I2>(
-        mut pattern: I1,
-        mut code: I2,
-        placeholders: &[Var],
-        match_: Match,
-    ) -> Option<Match>
-    where
-        T: AstNode,
-        I1: Iterator<Item = T>,
-        I2: Iterator<Item = T>,
-    {
-        pattern
-            .by_ref()
-            .zip(code.by_ref())
-            .fold(Some(match_), |accum, (a, b)| {
-                accum.and_then(|match_| {
-                    check(
-                        &a.syntax().clone().into(),
-                        &b.syntax().clone().into(),
-                        placeholders,
-                        match_,
-                    )
-                })
+impl ra_parser::TreeSink for PatternTreeSink<'_> {
```

Comment:
Nifty! So we essentially re-use existing parser for parsing SSR templates.

I am not sure how I feel about this:

On the one hand, this is definitely the "proper" way to do this, and I imagine that, long term, it might unlock interesting features like determining the syntax category of the pattern based on the parser state. 

On the other hand, this exposes ssr to a low-level API, which is verbose and can potentially suffer breakage. (in particular, we currently don't handle hygiene and macro captures properly, and that might change the ideal interface here). 

My gut feeling is that, at the current, stage, the easier to maintain solution would be to hack-in the placeholder support based on string parsing. Basically, we replace placeholders with specially-crafted identifiers, and then look for magic idents in the parse tree. The main benefit I see is that this only needs `String -> SyntaxNode` API, and that is unlikely to be changed soon.

Reply:
Now no longer depends on the TreeSink and TokenSource APIs